File: plus_one.py
'''
Created on 2014-2-21

leetcode : Plus One

Problem : 

Given a non-negative number represented as an array of digits, plus one to the number.

The digits are stored such that the most significant digit is at the head of the list.

Solve : 

@author: xqk
'''


class Solution:
    # @param digits, a list of integer digits
    # @return a list of integer digits
    def plusOne(self, digits):
        l = len(digits)
        flag = 1
        for i in range(len(digits) ):
            pos = l - 1 - i ;
            if digits[pos] + flag == 10:
                digits[pos] = 0
                if pos == 0 :
                    digits = [1] + digits
            else:
                digits[pos] = digits[pos] + 1
                flag = 0
                break;
        return digits  
digits = [9,9]
test = Solution()
print test.plusOne(digits)

File: Cplusplus/Plus_One.cpp
/*
Leetcode : Plus_One.cpp
Given a non-negative number represented as an array of digits,

plus one to the number.

The digits are stored such that the most significant digit is

at the head of the list.
*/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<int> plusOne(vector<int> &digits) {
        int carry = 0;
        vector<int> result(digits.size()+1 , 0);
        int len = digits.size()-1;
        digits[len] += 1;

        for(int i = digits.size()-1 ; i>= 0 ; i--)
        {
            int tmp = digits[i] + carry;
            carry = tmp/10;

            if(carry)
                result[len-i] = tmp%10;
            else
                result[len-i] = tmp;

        }

        if(carry)
            result[len+1] = 1;
        else
            result.pop_back();
        reverse(result.begin() , result.end());
        return result;
    }
};

int main()
{
    Solution test = Solution();
    vector<int> n(1,9);
    vector<int> r = test.plusOne(n);
    for(int i = 0 ; i < r.size() ; i++)
        cout<<r[i];
}
